### PR TITLE
SQL Alchemy Results backend not working with results extended set to true

### LINKED ISSUE #5436: Celery 4.3 result_extended doesn't push extended meta data to some backends when set to true
# Checklist

- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**:

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
celery -A celeryacl report

software -> celery:4.3.0 (rhubarb) kombu:4.5.0 py:3.6.8
billiard:3.6.0.0 py-amqp:2.4.2
platform -> system:Linux arch:64bit
kernel version:4.18.0-1013-azure imp:CPython
loader -> celery.loaders.app.AppLoader
settings -> transport:amqp results:db+mysql+pymysql://celery:**@X.X.X.X:XXXX/celeryresults
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: N/A or Unknown
* **Minimal Celery Version**: N/A or Unknown
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
pip freeze
amqp==2.4.2
asn1crypto==0.24.0
bcrypt==3.1.6
billiard==3.6.0.0
celery==4.3.0
cffi==1.12.2
Click==7.0
cryptography==2.6.1
Flask==1.0.2
Flask-Cors==3.0.7
itsdangerous==1.1.0
Jinja2==2.10
kombu==4.5.0
MarkupSafe==1.1.1
mysql-connector-python==8.0.15
netmiko==2.3.3
paramiko==2.4.2
protobuf==3.7.1
pyasn1==0.4.5
pycparser==2.19
PyMySQL==0.9.3
PyNaCl==1.3.0
pyserial==3.4
pytz==2018.9
PyYAML==5.1
scp==0.13.2
six==1.12.0
SQLAlchemy==1.2.14
textfsm==0.4.1
vine==1.3.0
Werkzeug==0.15.2
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
```

</p>
</details>

# Expected Behavior

With the following configured "result_extended" to true I expected to see more meta data in the backend results DB but I don't. I just see the standard results.

app = Celery('tasks',
             backend='db+mysql+pymysql://XXX:XXX@X.X.X.X:XXXX/celeryresults',
             broker='amqp://networkautomation:B2p2k6nb!@X.X.X.X/celery',
             result_extended=True)
# Actual Behavior

### PR DESCRIPTION
**Pull request by mrkcmo**
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

## Description

Fixes #5436

### CODE REVIEW

**Comment by mrkcmo** [comment]
@othalla Here is the new PR with the rebase for issue #5436

**Comment by mrkcmo** [comment]
@othalla  @thedrow  I am running into an error with "if self.app.conf.find_value_for_key('extended', 'result'):" in the "celery/backends/database/models.py". I want to make sure I am following your best practices and don't want to set includes without some better insight into the code base. What would you recommend I do in order to be able to check the configuration for a configured key in this file? Any recommendations would be great. Thanks!

File: celery/backends/database/models.py [hunk 1]
```diff
@@ -30,18 +30,34 @@ class Task(ResultModelBase):
...
+        results = {
             'task_id': self.task_id,
             'status': self.status,
             'result': self.result,
             'traceback': self.traceback,
             'date_done': self.date_done,
         }
+        if self.app.conf.find_value_for_key('extended', 'result'):
+            results.update(
+                {'name': self.task},
```
Comment (thedrow) [inline celery/backends/database/models.py]: Why are you creating a dictionary for each key?

**Comment by adimux** [comment]
I came across this PR because I too discovered that the database back-end did not store task's metadata.
I am wondering if it would be better to store all meta-data in a JSON field `meta`?
Not doing so resolves the problem only partially. Only the "extended" result will be stored in the db, leaving out custom meta-data that tasks might store.

File: celery/backends/database/models.py [hunk 2]
```diff
@@ -30,18 +30,34 @@ class Task(ResultModelBase):
     date_done = sa.Column(sa.DateTime, default=datetime.utcnow,
                           onupdate=datetime.utcnow, nullable=True)
     traceback = sa.Column(sa.Text, nullable=True)
+    name = sa.Column(sa.String(155))
+    args = sa.Column(sa.String(155))
+    kwargs = sa.Column(sa.String(155))
```
Comment (adimux) [inline celery/backends/database/models.py]: Sorry for intruding, but I was wondering if there was a specific reason why the size is 155 for these fields?
Especially for the args / kwargs, which can be long.

Comment (mrkcmo) [reply celery/backends/database/models.py]: > 
> 
> Sorry for intruding, but I was wondering if there was a specific reason why the size is 155 for these fields?
> Especially for the args / kwargs, which can be long.

No real reason. In my implementation those aren't large fields. If we need them to be larger than 155 characters then we can do that.

File: celery/backends/database/models.py (hunk 1 above)
Comment (mrkcmo) [reply celery/backends/database/models.py]: @thedrow I didn't mean to lol. I just meant to add them to the existing dict. I will fix that.

Comment (mrkcmo) [reply celery/backends/database/models.py]: Would this work better?
```
        if self.app.conf.find_value_for_key('extended', 'result'):
            results.update(
                {
                    'name': self.task,
                    'args': self.args,
                    'kwargs': self.kwargs,
                    'worker': self.worker,
                    'retries': self.retries,
                    'queue': self.queue,
                }
            )
```

File: celery/backends/database/models.py (hunk 2 above)
Comment (thedrow) [reply celery/backends/database/models.py]: Make them text fields instead.
The name can be more limited.

File: celery/backends/database/models.py (hunk 1 above)
Comment (thedrow) [reply celery/backends/database/models.py]: Yes that should work better.

File: celery/backends/database/models.py [hunk 3]
```diff
@@ -30,18 +30,36 @@ class Task(ResultModelBase):
...
     def to_dict(self):
-        return {
+        results = {
             'task_id': self.task_id,
             'status': self.status,
             'result': self.result,
             'traceback': self.traceback,
             'date_done': self.date_done,
         }
+        if self.app.conf.find_value_for_key('extended', 'result'):
```
Comment (thedrow) [inline celery/backends/database/models.py]: Seems like you are accessing the app where it is not available.
You should probably require it as an argument to `__init__()`.

File: celery/backends/database/models.py (hunk 2 above)
Comment (mrkcmo) [reply celery/backends/database/models.py]: Changed them to type Text and set them so they could be nullable.

File: celery/backends/database/__init__.py [hunk 4]
```diff
@@ -100,7 +100,7 @@ def ResultSession(self, session_manager=SessionManager()):
 
     @retry
     def _store_result(self, task_id, result, state,
-                      traceback=None, max_retries=3, **kwargs):
+                      args, worker, retries, queue, traceback=None, max_retries=3, **kwargs):
```
Comment (adimux) [inline celery/backends/database/__init__.py]: If am not mistaken, these arguments are not passed to _store_result.
They are part of the `request` object that is passed as an argument.
I tried this code locally and it only worked when I used the request to retrieve this data.

File: celery/backends/database/__init__.py [hunk 5]
```diff
@@ -113,6 +113,13 @@ def _store_result(self, task_id, result, state,
             task.result = result
             task.status = state
             task.traceback = traceback
+            if self.app.conf.find_value_for_key('extended', 'result'):
+                task.name = task
+                task.args = args
```
Comment (adimux) [inline celery/backends/database/__init__.py]: since args and kwargs are tuples, dicts, maybe we should consider storing them in JSON?

Comment (adimux) [reply celery/backends/database/__init__.py]: my source is https://github.com/celery/celery/blob/master/celery/backends/base.py#L701

File: celery/backends/database/__init__.py (hunk 4 above)
Comment (adimux) [reply celery/backends/database/__init__.py]: source: my source is https://github.com/celery/celery/blob/master/celery/backends/base.py#L701

File: celery/backends/database/__init__.py (hunk 5 above)
Comment (adimux) [reply celery/backends/database/__init__.py]: (or maybe using the result serializer `self.serializer` to encode the args/kwargs)

Comment (adimux) [reply celery/backends/database/__init__.py]: Btw: I can help if you don't have time.

File: celery/backends/database/__init__.py (hunk 4 above)
Comment (mrkcmo) [reply celery/backends/database/__init__.py]: > 
> 
> source:https://github.com/celery/celery/blob/master/celery/backends/base.py#L701

I would happy accept any help with making this work. I am not familiar enough with the app to know what I am doing at this point. lol

File: celery/backends/database/__init__.py (hunk 5 above)
Comment (mrkcmo) [reply celery/backends/database/__init__.py]: @adimux I would happy accept some help!

Comment (mrkcmo) [reply celery/backends/database/__init__.py]: @adimux added you as a collab on my fork: https://github.com/mrkcmo/celery/invitations

Comment (adimux) [reply celery/backends/database/__init__.py]: great, working on it

Comment (mrkcmo) [reply celery/backends/database/__init__.py]: @adimux  any updates?

Comment (adimux) [reply celery/backends/database/__init__.py]: I couldn't work on it a lot the last weeks, but this week I'll probably have a working version.

File: celery/backends/cassandra.py [hunk 6]
```diff
@@ -8,7 +8,7 @@
 from celery.exceptions import ImproperlyConfigured
 from celery.utils.log import get_logger
 
-from .base import BaseBackend
+from .base import StructuredStoreBackend
```
Comment (adimux) [inline celery/backends/cassandra.py]: How do I test that the Cassandra back-end actually works?
I really don't have time to do a manual test and the unit tests don't actually test with a live Cassandra database unlike the SQLAlchemy back-end.

**Comment by adimux** [comment]
@thedrow We have a working version now. I also tried to fix Cassandra and AMQP though I couldn't test Cassandra and the unit tests are not using a live db.

**Comment by auvipy** [comment]
check the trivis

File: celery/backends/database/models.py (hunk 3 above)
Comment (mrkcmo) [reply celery/backends/database/models.py]: @adimux took over the implementation and has fixed this issue.

File: celery/backends/database/__init__.py (hunk 4 above)
Comment (mrkcmo) [reply celery/backends/database/__init__.py]: Resolved by @adimux

File: celery/backends/database/__init__.py (hunk 5 above)
Comment (mrkcmo) [reply celery/backends/database/__init__.py]: Resolved by @adimux

**Comment by mrkcmo** [comment]
@auvipy it is showing in travis the following flake 8 error:
 
`t/unit/backends/test_base.py:449:1: W293 blank line contains whitespace`

But there is no whitespace. I have checked the file locally and on Github. Not sure why travis is throwing this error.

**Comment by adimux** [comment]
By the way, the UTs do not use a live Cassandra database and I didn't test manually with one.
How do you test it usually? @auvipy

**Comment by auvipy** [comment]
It would be better to create separate PR for each back end for easier review.

**Comment by adimux** [comment]
@auvipy The thing is, I created a different base shared by the 3 back-end to avoid repeating some code related to extracting the task "meta".

**Comment by auvipy** [comment]
the base should work differently for each back end. If not, you need to modify it

**Comment by adimux** [comment]
What do you mean by "should work differently for each back end"?

**Comment by mrkcmo** [comment]
@auvipy I would tend to agree with @adimux  in that the base structure of the meta data gathering and structuring can be at the base but obviously you need separate backend processing of that data which already exists. What do we hope to accomplish if we split the base by backend type?

**Comment by auvipy** [comment]
The base should work separately for all back end. And changes should be made for each back end in separate PR.

**Comment by adimux** [comment]
Here is a PR for the AMQP one https://github.com/celery/celery/pull/5495

**Comment by adimux** [comment]
> @auvipy it is showing in travis the following flake 8 error:
> 
> `t/unit/backends/test_base.py:449:1: W293 blank line contains whitespace`
> 
> But there is no whitespace. I have checked the file locally and on Github. Not sure why travis is throwing this error.

I don't understand why it's saying that either. @auvipy Any idea?

**Comment by auvipy** [comment]
I do face this type of behaviour at some time. flake8 bug

**Comment by mrkcmo** [comment]
Please disregard I will close out this pull request and create a new fork to work on this issue.

**Comment by adimux** [comment]
@mrkcmo I modified this PR so that it fixes SQLAlchemy. And I created the other PR to fix AMQP. What's the issue?

**Comment by mrkcmo** [comment]
Sorry I didn’t release you put the fix in for SQL Alchemy. Did it pass travis?

Chris Fortner

________________________________

**Comment by adimux** [comment]
Yes, except for the pyflake check, which is apparently a bug as @auvipy said.

**Comment by mrkcmo** [comment]
@adimux  can you push those same fixes to this new repo?  https://github.com/mrkcmo/celery/invitations as I deleted the fork and just created a new one when they asked for the split.

**Comment by mrkcmo** [comment]
If you push the fixes here or potential to your own fork and then submit the PR that would be great. I kind of screwed up when I deleted the repo.

**Comment by mrkcmo** [comment]
I can still see the changes in the old (closed) PR….I can just make those same changes in the new repo and then resubmit the PR is you are busy. Just let me know!

Thanks so much for you help BTW!

Chris Fortner

________________________________

**Comment by adimux** [comment]
I am going to push to your fork

**Comment by adimux** [comment]
You can keep the same PR but just change the remote / branch. No need to create a new one.

**Comment by mrkcmo** [comment]
@adimux since I delete the original fork it doesn't allow me to reopen of change the source for this PR. Is there a trick you know to fix this? If not I can just do another PR with the same name.

**Pull request closed without merging**